src/backend/session/libseat.rs

```diff
@@ -0,0 +1,261 @@
+#[derive(Debug)]
+pub struct LibSeatSessionNotifier {
+    internal: Rc<LibSeatSessionImpl>,
+    signaler: Signaler<SessionSignal>,
+}
+
+impl LibSeatSession {
+    /// Tries to create a new session via libseat.
+    pub fn new<L>(logger: L) -> Result<(LibSeatSession, LibSeatSessionNotifier), Error>
+    where
+        L: Into<Option<::slog::Logger>>,
+    {
+        let logger = crate::slog_or_fallback(logger)
+            .new(o!("smithay_module" => "backend_session", "session_type" => "libseat"));
+
+        let active = Arc::new(AtomicBool::new(false));
+        let signaler = Signaler::new();
+
+        let seat = {
+            let enable = {
+                let active = active.clone();
+                let signaler = signaler.clone();
+                let logger = logger.clone();
+                move |_seat: &mut libseat::SeatRef| {
+                    debug!(logger, "Enable callback called");
+                    active.store(true, Ordering::SeqCst);
+                    signaler.signal(SessionSignal::ActivateSession);
+                }
+            };
+            let disable = {
+                let active = active.clone();
+                let signaler = signaler.clone();
+                let logger = logger.clone();
+                move |seat: &mut libseat::SeatRef| {
+                    debug!(logger, "Disable callback called");
+                    active.store(false, Ordering::SeqCst);
+                    seat.disable().unwrap();
+                    signaler.signal(SessionSignal::PauseSession);
+                }
+            };
+
+            Seat::open(enable, disable)
+        };
+
+        seat.map(|mut seat| {
+            // In some cases enable_seat event is avalible right after startup
+            // so, we can dispatch it
+            seat.dispatch(0).unwrap();
+
+            let seat_name = seat.name().to_owned();
+
+            let internal = Rc::new(LibSeatSessionImpl {
+                seat: RefCell::new(seat),
+                active,
+                devices: RefCell::new(HashMap::new()),
+                logger,
+            });
+
+            (
+                LibSeatSession {
+                    internal: Rc::downgrade(&internal),
+                    seat_name,
+                },
+                LibSeatSessionNotifier { internal, signaler },
+            )
+        })
+        .map_err(|_e| Error::FailedToOpenSession)
+    }
+}
+
+impl Session for LibSeatSession {
+    type Error = Error;
+
+    fn open(&mut self, path: &Path, _flags: OFlag) -> Result<RawFd, Self::Error> {
+        if let Some(session) = self.internal.upgrade() {
+            debug!(session.logger, "Opening device: {:?}", path);
+
+            session
+                .seat
+                .borrow_mut()
+                .open_device(&path)
+                .map(|(id, fd)| {
+                    session.devices.borrow_mut().insert(fd, id);
+                    fd
+                })
+                .map_err(|_| Error::Unknown)
```

Comment:
I think we'd prefer to forward the error, rather than completely drop it into an `Unknown` enum variant.

Reply:
https://github.com/Smithay/smithay/pull/292/commits/ace7a27d908b7ac7c21113ba2d60b08cdcb1dbe7
Now all methods have custom error types and are forwarding errno values properly.